module/zfs/dbuf.c

```diff
@@ -302,6 +303,12 @@ dbuf_hash(void *os, uint64_t obj, uint8_t lvl, uint64_t blkid)
 	return (cityhash4((uintptr_t)os, obj, (uint64_t)lvl, blkid));
 }
 
+#define	DBUF_STATE_CHANGE(db, op, state, why) do {		\
+	(db)->db_state op state;				\
+	DTRACE_PROBE2(dbuf__state_change, dmu_buf_impl_t *, db,	\
+	    const char *, why);					\
+} while (0)
```

Comment:
I presume this is done as a `#define` and not a `static inline` to preserve the `DTRACE_PROBE2` call location?  If that's the case, please add a comment to that effect above the macro definition. 

Additionally, the only operator (op) I see passed to this macro is `=`, can we change the macro to `DBUF_SET_STATE(db, state, why)` as a simplification and co locate it with the other `DBUF_*` macros in  `include/sys/dbuf.h`.

Reply:
db state becomes a bitmask, hence the operation.

Reply:
```#define	DTRACE_PROBE2(prov, name, arg1, arg2) {			\
	extern void __dtrace_##prov##___##name(unsigned long,	\
	    unsigned long);					\
	__dtrace_##prov##___##name((unsigned long)arg1,		\
	    (unsigned long)arg2);				\
}
```
It doesn't look like it actually uses the file / line info

Reply:
It would make sense to have it as a macro if we were logging the state before and after, but he wasn't actually doing that so I guess don't need it.

You'd rather see:
```
else if (db->db_state & (DB_READ|DB_PARTIAL))
		DBUF_STATE_CHANGE(db, |=, DB_FILL,
		    "notifying of followup partial fill");
	dbuf_dirty_leaf_common(&dds);
```

as:
```
else if (db->db_state & (DB_READ|DB_PARTIAL))
		dbuf_set_state(db, db->db_state |= DB_FILL,
		    "notifying of followup partial fill");
	dbuf_dirty_leaf_common(&dds);
```
That seems slightly more verbose but I'm fine with whatever.

Reply:
Unless there's a good reason for using a macro here, and I'm not sure there is, then yes let's implement this as an `inline` function.  Like so: (note the `|` instead of `|=`).

```c
	else if (db->db_state & (DB_READ | DB_PARTIAL)) {
		dbuf_set_state(db, db->db_state | DB_FILL,
		    "notifying of followup partial fill");
		...
	}
```

Reply:
Honestly, looking at this again.  I have to wonder if we actually want to use any helper function/macro here.  Directly calling the `DTRACE_PROBE2` isn't bad at all, and is very clear.  For example:

```diff
diff --git a/module/zfs/dbuf.c b/module/zfs/dbuf.c
index 0247860..b400fe2 100644
--- a/module/zfs/dbuf.c
+++ b/module/zfs/dbuf.c
@@ -1210,6 +1210,8 @@ dbuf_read_done(zio_t *zio, const zbookmark_phys_t *zb, con
                ASSERT(db->db_blkid != DMU_BONUS_BLKID);
                ASSERT3P(db->db_buf, ==, NULL);
                db->db_state = DB_UNCACHED;
+               DTRACE_PROBE2(dbuf__state_change, dmu_buf_impl_t *, db,
+                   const char *, "i/o error");
        } else if (db->db_level == 0 && db->db_freed_in_flight) {
                /* freed in flight */
                ASSERT(zio == NULL || zio->io_error == 0);
@@ -1219,11 +1221,15 @@ dbuf_read_done(zio_t *zio, const zbookmark_phys_t *zb, c
                db->db_freed_in_flight = FALSE;
                dbuf_set_data(db, buf);
                db->db_state = DB_CACHED;
+               DTRACE_PROBE2(dbuf__state_change, dmu_buf_impl_t *, db,
+                   const char *, "freed in flight");
        } else {
                /* success */
                ASSERT(zio == NULL || zio->io_error == 0);
                dbuf_set_data(db, buf);
                db->db_state = DB_CACHED;
+               DTRACE_PROBE2(dbuf__state_change, dmu_buf_impl_t *, db,
+                   const char *, "successful read");
        }
        cv_broadcast(&db->db_changed);
        dbuf_rele_and_unlock(db, NULL, B_FALSE);
```

Unless there's strong objection, why don't we call the `DTRACE_PROBE` directory.  I'd also like to get @brad-lewis's thoughts on this since he recently fixed up the DTRACE_PROBE on Linux.

Reply:
I don't have strong feelings about use of a dbuf_set macro or not.  I would have probably called the DTRACE_PROBE2() directly as Brian suggested.  It's a litter clearer to me and the macro doesn't save you that much.

Reply:
As you wish. I prefer a single line, but don't actually care. I'll accept a quorum of 2.

Reply:
Actually ... I'm sorry but declaring the argument types over and over is just gross.
Can we at least do this?
```
#define	DTRACE_SET_STATE(db, why) \
	DTRACE_PROBE2(dbuf__state_change, dmu_buf_impl_t *, db,	\
	    const char *, why)
...
	if (buf == NULL) {
		/* i/o error */
		ASSERT(zio == NULL || zio->io_error != 0);
		ASSERT(db->db_blkid != DMU_BONUS_BLKID);
		ASSERT3P(db->db_buf, ==, NULL);
		db->db_state = DB_UNCACHED;
		DTRACE_SET_STATE(db, "i/o error");
	} else if (db->db_level == 0 && db->db_freed_in_flight) {
		/* freed in flight */
		ASSERT(zio == NULL || zio->io_error == 0);
		arc_release(buf, db);
		bzero(buf->b_data, db->db.db_size);
		arc_buf_freeze(buf);
		db->db_freed_in_flight = FALSE;
		dbuf_set_data(db, buf);
		db->db_state = DB_CACHED;
		DTRACE_SET_STATE(db, "freed in flight");
	} else {
		/* success */
		ASSERT(zio == NULL || zio->io_error == 0);
		dbuf_set_data(db, buf);
		db->db_state = DB_CACHED;
		DTRACE_SET_STATE(db, "successful read");
...
```

Reply:
Sure, I think that's a nice compromise.